### DISCUSSION TITLE
Using django-filter with class-based views

### DISCUSSION THREAD

**Discussion by perlun** [carltongibson/django-filter#1369] [Q&A] [answered]
Hi,

The docs are pretty clear on how to use this library with traditional Django views:

```python
def product_list(request):
    filter = ProductFilter(request.GET, queryset=Product.objects.all())
    return render(request, 'my_app/template.html', {'filter': filter})
```

_However_, we are not using function-based views in our application but [class-based views](https://docs.djangoproject.com/en/3.2/topics/class-based-views/intro/). I have tried various approaches on trying to get this working but I really don't get it.

Here's my last attempt:

```python
class CustomersView(LoginRequiredMixin):
    queryset = Customer.objects.all()

    filter = CustomerFilter(self.request, queryset=self.queryset)
```

Obviously, this can't work since `filter = CustomerFilter(request, queryset=queryset)` relies on things from `self` which isn't available when declaring properties like that... :slightly_smiling_face: 

I've also tried various approaches which involved overriding `get_queryset`, like this:

```python
    def get_queryset(self):
        filter = CustomerFilter(self.request, queryset=self.queryset)
        return self.queryset.filter(filter)
```

But with this I'm getting an error `cannot unpack non-iterable CustomerFilter object`. There's probably something obvious I'm missing out here... please let me know what it is. :smile:

**Comment by carltongibson** [comment, marked as answer]
Take a look at the bundled generic view implementation: 

https://github.com/carltongibson/django-filter/blob/2b2c13dbd65ed0f794a06b343b7c4793042caf13/django_filters/views.py#L76-L88

It should give you the guidance you need. 

In general you access the filtered QuerySet via `filter.qs`.

**Reply by perlun** [reply]
Thanks @carltongibson. I think I might have stumbled across that method actually, but I assumed there was some simpler way to accomplish this... But great, I'll give this a look. :+1: Thanks a lot!

(and btw, sorry for not making this a discussion thread in the first place. :man_facepalming:)

**Reply by perlun** [reply]
For reference, here's what I ended up with. Posting it here in case it helps others:

```python
    def get(self, request, *args, **kwargs):
        filterset = CustomerFilter(request.GET, queryset=self.queryset)

        self.object_list = filterset.qs
        context = self.get_context_data(filter=filterset,
                                        object_list=self.object_list)

        return self.render_to_response(context)
```

@carltongibson - maybe it would be worth adding this to the docs, perhaps somewhere close to https://django-filter.readthedocs.io/en/main/guide/usage.html#the-view? If you agree it's a good idea, I can fix a PR for this.
